clients/src/test/java/org/apache/kafka/clients/consumer/CooperativeStickyAssignorTest.java

```diff
@@ -78,6 +80,26 @@ public void testDecodeGeneration() {
         assertFalse(((CooperativeStickyAssignor) assignor).memberData(subscription).generation.isPresent());
     }
 
+    @Test
+    public void testCooperativeStickyAssignorHonorSubscriptionUserdataIfNoGenerationIdInField() {
+        Map<String, Integer> partitionsPerTopic = new HashMap<>();
+        partitionsPerTopic.put(topic, 2);
+        int higherGenerationId = 2;
+        int lowerGenerationId = 1;
+
+        assignor.onAssignment(new ConsumerPartitionAssignor.Assignment(partitions(tp1)), new ConsumerGroupMetadata(groupId, higherGenerationId, consumer1, Optional.empty()));
+        ByteBuffer userDataWithHigherGenerationId = assignor.subscriptionUserData(new HashSet<>(topics(topic)));
```

Comment:
All the other tests rely on `buildSubscriptionWithGeneration` and `buildSubscription`, I wonder if we should do the same here. Continuing on my previous comment, we could perhaps extend or add new versions of those to differentiate the old way from the new way.

Reply:
We can use `buildSubscriptionWithGeneration` for consumer2 here, but not for consumer1. Please note, the consumer1 provides different owned partitions and generation id as the one provided in subscription. I'm trying to test which data we should honor in different cases. I've updated the comment to make it clear.

Reply:
I think that your goal is to basically ensure that `findOwnedPartitionsWithGeneration` works as expected given all the possible inputs. Should we just add a unit test for this method for both assignors with all the possible inputs? The intent would be clearer in my opinion.

Reply:
You're right. Change to test `memberDataFromSubscription`